packages/hardhat-core/src/internal/core/providers/lazy-initialization.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  EthereumProvider,
+  JsonRpcRequest,
+  JsonRpcResponse,
+  RequestArguments,
+  ProviderFactory,
+} from "../../../types";
+import { CustomError } from "../errors";
+
+/**
+ * A class that delays the (async) creation of its internal provider until the first call
+ * to a JSON RPC method via request/send/sendAsync.
+ * Trying to use the EventEmitter API without calling request first (initializing the provider)
+ * will throw.
+ */
+export class LazyInitializationProvider implements EthereumProvider {
```

Comment:
This class re-implements the methods from `EventEmitter` like `EventEmitterWrapper` does. This could be done automagically by maybe looping through `Object.keys(EventEmitter.prototype)`.

I prefer this more explicit solution but we can discuss it further!

Reply:
I also much prefer the explicit solution